feat(messaging): implement messaging api

jharvey10 opened the pull request:
- Assorted repo config and CI updates
- Update run-mode to switch to DEV and PROD as only modes

BREAKING CHANGE:

- run-mode API has changed.
- TEST is no longer a mode
- PRODUCTION is now called PROD

jharvey10 committed:
feat(messaging): implement messaging api

- Aassorted repo config and CI updates
- Update run-mode to switch to DEV and PROD as only modes

BREAKING CHANGE:

- run-mode API has changed.
- TEST is no longer a mode
- PRODUCTION is now called PROD

jharvey10 committed:
refactor(messaging): convert callback registry to map

 - Update null usages to typescript optionals

francinelucca commented on package.json:
```diff
@@ -38,6 +38,8 @@
     "@types/jest": "^27.4.0",
     "@typescript-eslint/eslint-plugin": "^5.9.0",
+    "commander": "^8.3.0",
+    "dotenv": "^10.0.0",
```
are we using this package anywhere?

jdharvey-ibm replied:
I don't think so. I'll remove that.

francinelucca commented on packages/api/src/main/messaging/MessagingClient.ts:
```diff
@@ -0,0 +1,184 @@
+    this.replyCallbacks = new Map()
+  }
+
+  public async connect(retry = true): Promise<amqp.ConfirmChannel> {
```
should we add Jsdocs to the public functions here?

jdharvey-ibm replied:
I did a docs run through all of these PRs. Definitely let me know if you have any comments on them or things you think I should add/change

francinelucca replied:
LGTM!

francinelucca commented on packages/api/src/main/messaging/MessagingClient.ts:
```diff
@@ -0,0 +1,184 @@
+import { EventMessage, QueryMessage } from './config'
+import { MESSAGE_QUEUE_URL } from './constants'
+
+const RANDOM_QUEUE_NAME = ''
+const DEFAULT_ROUTING_KEY = ''
```
are these values meant to be empty?

jdharvey-ibm replied:
Yeah. I added some docs to these to clarify.

francinelucca commented on packages/api/src/main/messaging/config.ts:
```diff
@@ -0,0 +1,38 @@
+ * Each service gets a queue. Services bind their queue to one or more exchanges to receive
+ * messages.
+ */
+enum Queue {
+  Github = 'q_github_v1',
+  Logging = 'q_logging_v1',
+  Search = 'q_search_v1'
+}
```
are we using this anywhere?

jdharvey-ibm replied:
Not in this PR, but the logging one is used over [here](https://github.com/carbon-design-system/carbon-platform/blob/9ca0958b008ed2f72d573834971d6d87474495bb/services/logging/src/main/index.ts)

francinelucca commented on packages/api/src/main/messaging/constants.ts:
```diff
@@ -0,0 +1,41 @@
+}
+const DEFAULT_EXCHANGE_TYPE = 'fanout'
+
+// Message queue URL checking
+if (getRunMode() === PROD && !process.env.CARBON_MESSAGE_QUEUE_URL) {
+  throw new Error('CARBON_MESSAGE_QUEUE_URL must be set in PROD run mode')
+}
+if (MESSAGE_QUEUE_URL === LOCAL_MESSAGE_QUEUE_URL) {
+  console.warn(`${chalk.bgYellowBright.black('WARN')} Using localhost message queue url`)
+}
```
is the constants file the correct place to do these checks? should we check the env vars in the constants file by standard from now on?

jdharvey-ibm replied:
Definitely not where we want to do these, haha. I'm basically just waiting on the stuff you have on your branch to make it in and we can do a pass through all of these and make them use the new way of orchestrating them.

jdharvey-ibm replied:
Created issue #202

francinelucca commented on packages/api/src/main/messaging/constants.ts:
```diff
@@ -0,0 +1,41 @@
+  console.warn(`${chalk.bgYellowBright.black('WARN')} Using localhost message queue url`)
+}
+
+export {
+  DEBUG,
+  DEFAULT_BIND_PATTERN,
+  DEFAULT_EXCHANGE_OPTIONS,
+  DEFAULT_EXCHANGE_TYPE,
+  DEFAULT_QUEUE_OPTIONS,
+  MESSAGE_QUEUE_URL
+}
```
seems like most of these aren't referenced anywhere, are they part of a separate PR?

jdharvey-ibm replied:
They're used over in PR #193

francinelucca commented on packages/api/src/test/messaging/MessagingClient.test.ts:
```diff
@@ -0,0 +1,157 @@
+    await expect(client.connect(false)).rejects.toThrow()
+  })
+
+  it('quietly ignores replies for unknown correlation IDs', async () => {
```
wouldn't this test pass whether the correlation id was found or not? could we assert that replyCallbacks.delete wasn't called to ensure the function returned earlier?

jdharvey-ibm replied:
It would, yeah. The only purpose of this test is to ensure that a known non-existant value is handled gracefully. I'll add a comment to the test to clarify.

jharvey10 committed:
docs(messaging): add code and markdown docs

jharvey10 committed:
docs(api): correct run-mode docs

jharvey10 committed:
fix(messaging): address review comments

francinelucca commented on docs/packages-api.md:
```diff
@@ -18,14 +18,17 @@ messages to be logged.
+
+If the environment variable is not set, the mode will default to `DEV`.
+
+Settings this envvar to any other value will result in an error being thrown.
```
```suggestion
Setting this envvar to any other value will result in an error being thrown.
```

jharvey10 committed:
Update docs/packages-api.md

Co-authored-by: Francine Lucca <40550942+francinelucca@users.noreply.github.com>

francinelucca commented on .eslintrc.js:
```diff
@@ -69,6 +69,13 @@ module.exports = {
           }
         ]
       }
+    },
+    {
+      files: ['*.ts'],
+      rules: {
+        'no-unused-vars': 'off',
+        '@typescript-eslint/no-unused-vars': ['error']
+      }
```
could the services/web-app/lib/file-cache.js sonarcloud error be caused by this change?

jdharvey-ibm replied:
Nah, that was just me being dumb and editing a change on the github UI rather than in my IDE. I commented out the only use of a var, but didn't comment out the var definition itself.

This change is needed because the baseline eslint rule for unused vars doesn't properly handle typescript types and a couple of other weird edge cases.

jharvey10 committed:
fix(messaging): connect race condition

- Fixes a bug which allowed the messaging client singleton to establish multiple connections to the
  message broker at the same time.

jharvey10 committed:
refactor(messaging): remove async from client connect

Async was only used to wrap one value in a promise.
Now using Promise.resolve() instead.

jharvey10 committed:
refactor(messaging): address code review comments

- remove need for eslint ignores and async promise executors
- Thanks for the suggestions, Francine!

francinelucca reviewed (dismissed):
LGTM!

jharvey10 committed:
Merge branch 'main' of github.com:carbon-design-system/carbon-platform into feat/messaging

mattrosno commented on docs/packages-api.md:
```diff
@@ -18,14 +18,17 @@ messages to be logged.
 
 - `DEV`
-- `PRODUCTION`
+- `PROD`
```
Since we're now using this in the web-app, can you update from PRODUCTION to PROD in:

- [ ] `constants.js:7`
- [ ] `constants.js:9`

jdharvey-ibm replied:
Done!

mattrosno reviewed (dismissed):
Thanks for the thorough documentation! I just read through the markdown.

jharvey10 committed:
Merge branch 'main' of github.com:carbon-design-system/carbon-platform into feat/messaging

jharvey10 committed:
refactor: update run mode uses

convery PRODUCTION to PROD

francinelucca reviewed (approved):
LGTM
